samcli/commands/local/lib/local_lambda.py

```diff
@@ -203,13 +207,10 @@ def get_aws_creds(self):
         result = {}
 
         # to pass command line arguments for region & profile to setup boto3 default session
-        if boto3.DEFAULT_SESSION:
-            session = boto3.DEFAULT_SESSION
-        else:
-            session = boto3.session.Session()
-
-        profile_name = session.profile_name if session else None
-        LOG.debug("Loading AWS credentials from session with profile '%s'", profile_name)
+        LOG.debug("Loading AWS credentials from session with profile '%s'", self.aws_profile)
+        # boto3.session.Session is not thread safe. To ensure we do not run into a race condition with start-lambda
+        # or start-api, we create the session object here on every invoke.
+        session = boto3.session.Session(profile_name=self.aws_profile, region_name=self.aws_region)
```

Comment:
For the customer, it really doesn't matter. I don't think a debug statement for this makes sense. 

For further context, this is how we did it prior to the work we did on Layers. We got fancy in using the default session. Turns out that only really worked because we were calling `boto3.client`. Once we moved this call to be computed, I was seeing sporadic failures on a `KeyError` within `botocore`. So I reverted us back to our previous state, which ensures Sessions get created per invoke. Which means for `start-lambda` and `start-api` that could be per request.